core/src/main/scala/org/apache/spark/TaskEndReason.scala

```diff
@@ -81,7 +81,7 @@ case object Resubmitted extends TaskFailedReason {
  */
 @DeveloperApi
 case class FetchFailed(
-    bmAddress: BlockManagerId,  // Note that bmAddress can be null
+    bmAddress: Location,  // Note that bmAddress can be null
```

Comment:
This is a backwardly incompatible change, which also impacts event files.
Unfortunately it looks like most folks who worked on this in past are no longer active.
+CC @tgravescs, @dongjoon-hyun for additional review.

Reply:
For the event files, we still log the json of `BlockManagerId`. So I think it's still compatible, no?

Besides, do you worry about the possible backward incompatibility due to the reference out of Spark?

Reply:
There are two changes here: a) change to public api (programmatic, REST api, etc), b) change to generated/persisted data.
Both need to be addressed.

Specifically about (b), ​with `bmAddress` changing to `Location` - since it need not be a `BlockManagerId` anymore - json serde needs to account for it. Currently, it is a TODO [here](https://github.com/apache/spark/pull/31876/files#diff-d87296ea32a40cc886384a0afaa3396ab8e68a5bbf9c05e399dc91fc86a7d7c4).

Spark history server, REST consumers, other apps depending on event files - will need a way to identify Location type and serde all valid Location types.

Reply:
> a) change to public api (programmatic, REST api, etc)

Technically, it's really a problem. Although, I can't imagine how users would use it as an API.

And I have a new idea that we can introduce a new fetch failed class for the custom location and leave this one unchanged. For example, we can have `CustomStorageFetchFailed`. Thus, we the location is `BlockManagerId ` then we use `FetchFailed`, otherwise, uses `CustomStorageFetchFailed`.  WDYT? 

> b) change to generated/persisted data.

I think we won't change the data of `BlockManagerId` here. If we find the `Location` is a `BlockManagerId`, we'd still output as `("Block Manager Address" -> blockManagerAddress)`.  So, IIUC, it won't cause problems.

The only problem is the custom location. It's new data, e.g., `("XXXLocation" -> XXXLocationJson)`. So it can be a problem if users use the old version Spark to load event files. Although, I think this's really an unexpected usage.

Reply:
> And I have a new idea that we can introduce a new fetch failed class for the custom location and leave this one unchanged. For example, we can have CustomStorageFetchFailed. Thus, we the location is BlockManagerId then we use FetchFailed, otherwise, uses CustomStorageFetchFailed. WDYT?

`CustomStorageFetchFailed` looks like a promising approach, we will need to think through what the implications of it would be would on the face of it, it should address immediate concerns IMO.
Thoughts @attilapiros, @tgravescs ?

> The only problem is the custom location. It's new data, e.g., ("XXXLocation" -> XXXLocationJson). So it can be a problem if users use the old version Spark to load event files. Although, I think this's really an unexpected usage.

There are couple of issues here:
* A simpler question of how to handle custom location - from programmatic and data point of view.
* How to handle different shuffle impls being in play for the same event directory.
  * If deployments have multiple shuffle infra in use over course of time (or different clusters with different configs and a shared history event dir), each with their own Location's.
  * How will SHS/REST api, etc understand which location class is being used/how to parse them.

I actually dont have good solutions on this - other than adding some metadata per location record to indicate the 'type'.
Any other thoughts ?

Reply:
+1 for a new class without introducing a breaking change. Thank you for pinging me, @mridulm .

Reply:
I think we'll have to add interfaces to `Location` in order to support json serde , e.g.,

```java

public JValue serializeToJson();

public Location deserilaizeFromJson(json: JValue);

```

> How to handle different shuffle impls being in play for the same event directory.

Adding metadata is good idea, we can have the format like,


```json
"Mapstatus Location": {
  "type":  "xx.yy.zz", // qualified class name
  "content": { // content is generate by `Location.serializeToJson`
    "aaa":"bbb"
   }
}
```

with the constant format, end-users and SHS are able to parse them as well.

(I had an idea about SHS is to add the location type as the extension of the event log file. That's the way what compression does now.  But I think it doesn't solve the problem of REST case.)

BTW, I have added https://issues.apache.org/jira/browse/SPARK-35188 for this support.